component/publisher-client/src/main/java/org/wso2/am/analytics/publisher/client/MoesifClient.java

```diff
@@ -0,0 +1,206 @@
+            // put the elements in queue missed to eventQueue.
+            missedEventHandler.putMissedEvent(builder);
+            return;
+        }
+
+        // init moesif api client
+        MoesifAPIClient client = keyRetriever.getMoesifClient(moesifKey);
+        APIController api = client.getAPI();
+
+        APICallBack<HttpResponse> callBack = new APICallBack<HttpResponse>() {
+            public void onSuccess(HttpContext context, HttpResponse response) {
+                int statusCode = context.getResponse().getStatusCode();
+                if (statusCode == 200) {
+                    log.debug("Event successfully published.");
+                } else if (statusCode >= 400 && statusCode < 500) {
+                    log.error("Event publishing failed. Moesif returned " + statusCode);
+                } else {
+                    log.error("Event publishing failed.Retrying.");
+                    doRetry(orgId, builder);
+                }
+            }
+
+            public void onFailure(HttpContext context, Throwable error) {
+                int statusCode = context.getResponse().getStatusCode();
+
+                if (statusCode >= 400 && statusCode < 500) {
+                    log.error("Event publishing failed. Moesif returned:", statusCode);
+                } else if (error != null) {
+                    log.error("Event publishing failed.", error);
+                } else {
+                    log.error("Event publishing failed. Retrying.");
+                    doRetry(orgId, builder);
+                }
+
+            }
+        };
+        try {
+            api.createEventAsync(buildEventResponse(event), callBack);
+        } catch (IOException e) {
+            log.error("Analytics event sending failed. Event will be dropped", e);
+        }
+
+    }
+
+    private EventModel buildEventResponse(Map<String, Object> data) throws IOException, MetricReportingException {
+        Gson gson = new Gson();
+        String jsonString = gson.toJson(data);
+        String reqBody = jsonString.replaceAll("[\r\n]", "");
+
+        //      Preprocessing data
+        final URL uri = new URL((String) data.get(Constants.DESTINATION));
+        final String hostName = uri.getHost();
+
+        final String userIP = (String) data.get(Constants.USER_IP);
+
+        Map<String, String> reqHeaders = new HashMap<String, String>();
+
+        reqHeaders.put("User-Agent",
+                (String) data.getOrDefault(Constants.USER_AGENT_HEADER, Constants.UNKNOWN_VALUE));
+        reqHeaders.put("Content-Type", Constants.MOESIF_CONTENT_TYPE_HEADER);
+        reqHeaders.put("Host", hostName);
+
+        Map<String, String> rspHeaders = new HashMap<String, String>();
+
+        rspHeaders.put("Vary", "Accept-Encoding");
+        rspHeaders.put("Pragma", "no-cache");
+        rspHeaders.put("Expires", "-1");
+        rspHeaders.put("Content-Type", "application/json; charset=utf-8");
```

Comment:
Let's use constants as Strings like "Content-Type" are reused in many places.

Reply:
Ack.